keps/sig-node/4328-node-label-manager/README.md

```diff
@@ -0,0 +1,404 @@
+Every time a node's label is updated, the controller needs to iterate all pods assigned to the node. This may cause potential scaling issues.
+
+## Design Details
+
+Add a node label manager similar to the existing taint_eviction controller, but under the node lifecycle controller.
+
+The manager do the following things:
+
+- Listening to the changes of node labels
+- Iterating over all pods assigned to the nodes, check the NodeAffinity field, if `RequiredDuringSchedulingRequiredDuringExecution` exists, check if the selector still met to the new node label. If not, evict the pod.
+
+As for API definition, `RequiredDuringSchedulingRequiredDuringExecution` will have the same item as `requiredDuringSchedulingIgnoredDuringExecution`.
+
+```go
+type NodeAffinity struct {
+	RequiredDuringSchedulingRequiredDuringExecution *NodeSelector `json:"requiredDuringSchedulingRequiredDuringExecution,omitempty"`
+}
+```
+
+And NodeSelector already exists:
+```go
+type NodeSelector struct {
+	// Required. A list of node selector terms. The terms are ORed.
+	NodeSelectorTerms []NodeSelectorTerm `json:"nodeSelectorTerms" protobuf:"bytes,1,rep,name=nodeSelectorTerms"`
+}
+```
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+<!--
+Based on reviewers feedback describe what additional tests need to be added prior
+implementing this enhancement to ensure the enhancements have also solid foundations.
+-->
+
+##### Unit tests
+
+- Test that the validation returns an error when given an invalid NodeSelector.
+- Test that the validation pass when given a valid NodeSelector.
+- Test that the node_affinity plugin in scheduler handles the new type correctly.
+- Test that the NodeAffinityController receive the pod/node events, and handle them correctly(delete events occurred if pod's affinity doesn't match with node anymore). 
+
+##### Integration tests
+- Test that pods with `RequiredDuringSchedulingRequiredDuringExecution` node affinity are correctly scheduled.
+- Test that `RequiredDuringSchedulingRequiredDuringExecution` works well with other scheduler plugins.
+
+##### e2e tests
+- Test that pods that don't satisfy `RequiredDuringSchedulingRequiredDuringExecution` rule are correctly evicted.
+- Test that pods with `RequiredDuringSchedulingRequiredDuringExecution` are scheduled to correct nodes. 
```

Comment:
I would make it one test. Maybe like this:

- create a node that has a label
- create a pod that matches the label and has RequiredDuringExecution
- remove the node label, check that pod is evicted
- create another pod, check that it doesn't schedule.